envoy/ssl/handshaker.h

```diff
@@ -150,5 +159,94 @@ class HandshakerFactory : public Config::TypedFactory {
   virtual SslCtxCb sslctxCb(HandshakerFactoryContext& handshaker_factory_context) const PURE;
 };
 
+enum class SelectionResult {
+  // continue the TLS handshake.
+  Continue,
+  // block the TLS handshake.
+  Stop,
+  // terminate the TLS handshake.
+  Terminate
+};
+
+/**
+ * Used to return the result from an synchronous/asynchronous cert selection.
+ */
+class CertSelectionCallback {
+public:
+  virtual ~CertSelectionCallback() = default;
+
+  virtual Event::Dispatcher& dispatcher() PURE;
+
+  /**
+   * Called when the cert selection completes.
+   */
+  virtual void onCertSelectionResult(SSL_CTX* ctx) PURE;
```

Comment:
Should this take a `TlsContext&` instead of `SSL_CTX*`?

Reply:
When selection failed, people could `onCertSelectionResult(nullptr)`, it could be easy to use, while `TlsContext&` can not do it, thoughts?